### PR TITLE
consensus: Move DoS into contract validators to allow variability of DoS based on context and further fixes to ValidateMRC

### PR DESCRIPTION
**Pull request by jamescowens**
This PR moves the assignment of DoS into the contract validation methods to allow different DoS's to be assigned based on different context/error conditions. This is used in ValidateMRC because different validation failures have different severities.

ValidateMRC was also refactored to implement the DoS scoring, and several other problems corrected.

Note the conditions used for the testnet and mainnet for the exertion of the reject_payment_interval validation have been refined. They are compatible with all current blocks containing MRC on testnet.

### CODE REVIEW

File: src/gridcoin/contract/contract.h [hunk 1]
```diff
@@ -514,10 +514,11 @@ void ApplyContracts(
 //! \brief Perform contextual validation for the contracts in a transaction.
 //!
 //! \param tx Transaction to validate contracts for.
+//! \param DoS Mishavior score out parameter
```
Comment (barton2526) [inline src/gridcoin/contract/contract.h]: "Mishavior"

File: src/main.cpp [hunk 2]
```diff
@@ -395,10 +395,13 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CTransaction &tx, bool* pfMissingInput
...
 
     // Perform contextual validation for any contracts:
-    if (!tx.GetContracts().empty() && !GRC::ValidateContracts(tx)) {
-        return tx.DoS(25, error("%s: invalid contract in tx %s",
-            __func__,
-            tx.GetHash().ToString()));
+
+    int DoS = 0;
+    if (!tx.GetContracts().empty() && !GRC::ValidateContracts(tx, DoS)) {
+        return tx.DoS(DoS, error("%s: invalid contract in tx %s, assigning DoS mishehavior of %i",
```
Comment (barton2526) [inline src/main.cpp]: "mishehavior"

File: src/main.cpp [hunk 3]
```diff
@@ -1939,10 +1940,12 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck)
                     return false;
                 }
 
-                if (nVersion >= 11 && !GRC::ValidateContracts(tx)) {
-                    return tx.DoS(25, error("%s: invalid contract in tx %s",
-                        __func__,
-                        tx.GetHash().ToString()));
+                int DoS = 0;
+                if (nVersion >= 11 && !GRC::ValidateContracts(tx, DoS)) {
+                    return tx.DoS(DoS, error("%s: invalid contract in tx %s, assigning DoS mishehavior of %i",
```
Comment (barton2526) [inline src/main.cpp]: "mishehavior"

**Comment by jamescowens** [comment]
Interesting the linter didn't catch those.

**Review by div72 — APPROVED** [review]
Weak utACK.

**Comment by jamescowens** [comment]
We can address any issues you have in a true up.

**Pull request merged**